### DISCUSSION TITLE
How to measure total request duration?

### DISCUSSION THREAD

**Discussion by bartlomiejgawel** [justcoding121/Titanium-Web-Proxy#963] [Q&A] [answered]
I'm looking for a way to measure total request duration. It'll be useful to create metrics. Do you have any idea how to do that? I know I can measure timestamps inside BeforeRequest/BeforeRespone handlers but do you know cleaner approach how to measure it?

**Comment by justcoding121** [comment, marked as answer]
Yep ? there's built-in timing for this now. Flip on `EnableRequestTimingCapture` and read `e.Timing` in `AfterResponse`:

```csharp
proxyServer.EnableRequestTimingCapture = true;

proxyServer.AfterResponse += async (sender, e) =>
{
    var timing = e.Timing;
    if (timing == null) return;

    Console.WriteLine($"{e.HttpClient.Request.Url} took {timing.TotalDuration.TotalMilliseconds:F0}ms " +
                      $"(TTFB {timing.TimeToFirstByte?.TotalMilliseconds:F0}ms)");
};
```

You also get breakdowns like `TimeToFirstByte`, `ConnectionWaitDuration`, `ResponseDeliveryDuration`, etc. No need to roll your own timestamps in BeforeRequest/BeforeResponse.

Closing this out ? shout if you need anything else.
